### ISSUE TITLE
FYI: Are you sure to use "Gdk.SHIFT_MASK" in prefs.js

### ISSUE THREAD

**Issue by osamuaoki** [jqno/gnome-happy-appy-hotkey#1]
I have similar code and I found "Gdk.SHIFT_MASK" was undefined while debugging it under my environment of GNOME43.

In this code prefs.js file uses this. I think desired constant is "Gdk.ModifierType.SHIFT_MASK", instead.

You can test situation for your build environment using my test code under test-keys/* of https://github.com/osamuaoki/inputmethod-shortcuts

(I don't know if this issue is version dependent. I see similar code as yours every where. E.g., https://github.com/eonpatapon/gnome-shell-extension-caffeine/issues/279)

**Comment by jqno** [comment]
Hi,

Thanks for the heads-up. I'll admit that I looked very closely at how Caffeine implemented this, because I tried many things on my own and none of them seemed to work. This one does.

Looking at the documentation ( https://gjs-docs.gnome.org/gdk40~4.0/ ), indeed I don't see a `Gdk.SHIFT_MASK` constant, but I can't find `Gdk.ModifierType.SHIFT_MASK` there either.

I'm basically a newbie at this topic, so if you know of ways to improve this, I'm certainly interersted, especially because this code is very complicated.

**Comment by osamuaoki** [comment]
I am a newbie for javascript, too.  I never wrote javascript code even for web page... 

Yes, GTK4/gdk GJS documentation lacks Gdk.ModifierType.SHIFT_MASK.  But GTK4/gdk defines this as enum in C and introspection seems to pick it up.  So this is documentation generation code bug.  GTK3/gdk GJS doesn't suffer this problem.

I was in the same situation as you.  So I wrote a code test platform to get me out of these voodoo magic code lines.  This test code is a GNOME extension which presents us with `prefs.js` but no real code in `extension.js`.  It is residing in a sub-directory `test-keys/` of https://github.com/osamuaoki/inputmethod-shortcuts .  It has on/off-switch and keyval printing to button label.    Installing it should give you good idea how these keyval are in runtime.  Please play with it. It's self explanatory after using it.

Anyway, the most complicated lines are to avoid registering normal keyboard keys with or without Shift pressed.  These are "a"-"Z" "A"-"Z" "1"-"9", "0",  .... keys (plus the same keys' keyval under weird xkb such as Arabic, Russian ...).  The original code is in C found in gnome-control-panel so it is most tested.  But shortcut registered in there doesn't register `<Shift>Shift_R` or `<Control>Shift_L` or similar.  My latest GJS commit enables it.

The damage done by the error in your code is small.  It doesn't prevent user to register Shift-pressed normal keys such as "A".  This blacklist portion is mostly for SAFEGURD.  Not much damage.

Let me keep using the today's uploaded code for a while.  If they are good, I can write PR.

Until then, please look at my main code and test code.

**Comment by jqno** [comment]
Thanks for taking the time to explain everything! I've also followed the discussion in the Caffeine repo.

I tested it locally and I was indeed able to define hotkeys like `<Shift>a`. So I made the fix you suggested, and now it's no longer possible: good! See https://github.com/jqno/gnome-happy-appy-hotkey/commit/1c2149a66f75e6ba7b33ded621bfc8cfea057574
The new version is now pending review at extensions.gnome.org.

Thank you!
